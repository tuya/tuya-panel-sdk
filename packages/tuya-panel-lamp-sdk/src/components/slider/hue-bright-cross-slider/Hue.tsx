import React, { Component } from 'react';
import Color from 'color';
import { View, StyleSheet, Animated, Easing, StyleProp, TextStyle } from 'react-native';
import { LinearGradient, TYText } from 'tuya-panel-kit';
import { Svg, Path, Rect } from 'react-native-svg';
import ColorUtils from '../../../utils/color';

type Props = {
  /**
   * 文本样式
   */
  textStyle: StyleProp<TextStyle>;
  /**
   * 色相值
   */
  hue: number;
  /**
   * 尺寸
   */
  size: number;
  /**
   * 轨道宽度
   */
  trackWidth: number;
  /**
   * 标签文本
   */
  label: string;
};

export default class HueBrightSlider extends Component<Props> {
  constructor(props: Props) {
    super(props);

    this.animation.setValue(this.props.hue);
    this.animation.addListener(this.handleUpdateThumb);
  }

  // eslint-disable-next-line react/no-deprecated
  componentWillReceiveProps(nextProps: Props) {
    if (nextProps.hue !== this.props.hue) {
      this.needUpdate = true;
    }
  }

  componentDidUpdate() {
    if (this.needUpdate) {
      this.needUpdate = false;
      this.runAnimation(this.props.hue);
    }
  }

  private pathRef: Path;

  private animation: Animated.Value = new Animated.Value(0);

  private needUpdate = false;

  handleUpdateThumb = ({ value }: { value: number }) => {
    const color = ColorUtils.hsb2hex(value, 100, 100);
    const rgb = Color(color).rgbArray();
    this.pathRef &&
      this.pathRef.setNativeProps({ fill: [0, rgb[0] / 255, rgb[1] / 255, rgb[2] / 255, 1] });
  };

  runAnimation(value: number) {
    this.animation.stopAnimation();
    Animated.timing(this.animation, {
      toValue: value,
      duration: 16,
      easing: Easing.linear,
    }).start();
  }

  update(value: number) {
    this.runAnimation(value);
  }

  render() {
    const { size, hue, label, trackWidth, textStyle = { fontSize: 12 } } = this.props;
    const color = ColorUtils.hsv2rgba(hue, 1000, 1000);
    const { fontSize } = StyleSheet.flatten(textStyle);
    const lineHeight = fontSize * 1.5;
    return (
      <View style={{ width: trackWidth, height: size }}>
        <View
          style={{
            width: trackWidth,
            height: size,
            borderRadius: trackWidth / 2,
            overflow: 'hidden',
          }}
        >
          <LinearGradient
            gradientId="hue-bright-cross"
            x1="0%"
            x2="0%"
            y1="0%"
            y2="100%"
            stops={{
              '0%': '#FF0000',
              '12%': '#F8CB0E',
              '24%': '#80FE06',
              '36%': '#08FB2B',
              '49%': '#04FAFC',
              '60%': '#0243FC',
              '70%': '#8700F9',
              '82%': '#FC00EF',
              '92%': '#F00A5B',
              '100%': '#FF0000',
            }}
            style={{ width: trackWidth, height: size }}
          >
            <Rect x={0} y={0} width={trackWidth} height={size} />
          </LinearGradient>
        </View>
        <View
          style={[
            styles.text,
            {
              transform: [
                {
                  translateX: -size / 2,
                },
                {
                  translateY: -size / 2,
                },
                {
                  rotate: '-90deg',
                },
              ],
              height: lineHeight,
              width: size,
              left: trackWidth + lineHeight / 2,
              bottom: -lineHeight / 2,
            },
          ]}
        >
          <TYText style={[{ fontSize: 12 }, textStyle]}>{label}</TYText>
        </View>

        <Animated.View
          style={{
            position: 'absolute',
            right: trackWidth + 4,
            top: Animated.add(
              -8,
              this.animation.interpolate({
                inputRange: [0, 360],
                outputRange: [0, size],
              })
            ),
          }}
        >
          <Svg width="22" height="16">
            <Path
              ref={(ref: Path) => {
                this.pathRef = ref;
              }}
              d="M2,2L17,2L20,8L17,14L2,14Z"
              stroke="#fff"
              strokeWidth={2}
              fill={color}
              strokeLinecap="round"
              strokeLinejoin="round"
            />
          </Svg>
        </Animated.View>
      </View>
    );
  }
}
const styles = StyleSheet.create({
  text: {
    alignItems: 'flex-start',
    bottom: 0,
    justifyContent: 'center',
    position: 'absolute',
  },
});
