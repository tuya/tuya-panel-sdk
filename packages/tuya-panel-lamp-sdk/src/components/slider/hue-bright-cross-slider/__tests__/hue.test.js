import React from 'react';
import { shallow } from 'enzyme';
import Hue from '../Hue';

describe('Hue components', () => {
  it('basic render', () => {
    const wrapper = shallow(<Hue hue={30} />);
    expect(wrapper).toMatchSnapshot();
  });

  it('componentWillReceiveProps render', () => {
    let hue = 0;
    let size = 10;

    const wrapper = shallow(<Hue hue={hue} size={size} />);
    expect(wrapper).toMatchSnapshot();

    hue = 30;
    wrapper.setProps({ hue });

    size = 30;
    wrapper.setProps({ size });

    wrapper.instance().update(500);

    wrapper.instance().handleUpdateThumb({ value: 500 });

    wrapper.unmount();
  });
});
