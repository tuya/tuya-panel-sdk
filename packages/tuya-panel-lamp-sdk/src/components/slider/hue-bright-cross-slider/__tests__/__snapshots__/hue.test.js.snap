// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Hue components basic render 1`] = `
<View
  style={
    Object {
      "height": undefined,
      "width": undefined,
    }
  }
>
  <View
    style={
      Object {
        "borderRadius": NaN,
        "height": undefined,
        "overflow": "hidden",
        "width": undefined,
      }
    }
  >
    <LinearGradient
      gradientId="hue-bright-cross"
      stops={
        Object {
          "0%": "#FF0000",
          "100%": "#FF0000",
          "12%": "#F8CB0E",
          "24%": "#80FE06",
          "36%": "#08FB2B",
          "49%": "#04FAFC",
          "60%": "#0243FC",
          "70%": "#8700F9",
          "82%": "#FC00EF",
          "92%": "#F00A5B",
        }
      }
      style={
        Object {
          "height": undefined,
          "width": undefined,
        }
      }
      x1="0%"
      x2="0%"
      y1="0%"
      y2="100%"
    >
      <Rect
        height={0}
        rx={0}
        ry={0}
        width={0}
        x={0}
        y={0}
      />
    </LinearGradient>
  </View>
  <View
    style={
      Array [
        Object {
          "alignItems": "flex-start",
          "bottom": 0,
          "justifyContent": "center",
          "position": "absolute",
        },
        Object {
          "bottom": -9,
          "height": 18,
          "left": NaN,
          "transform": Array [
            Object {
              "translateX": NaN,
            },
            Object {
              "translateY": NaN,
            },
            Object {
              "rotate": "-90deg",
            },
          ],
          "width": undefined,
        },
      ]
    }
  >
    <TYText
      align={null}
      color={null}
      size={null}
      style={
        Array [
          Object {
            "fontSize": 12,
          },
          Object {
            "fontSize": 12,
          },
        ]
      }
      type={null}
      weight={null}
    />
  </View>
  <AnimatedComponent
    style={
      Object {
        "position": "absolute",
        "right": NaN,
        "top": NaN,
      }
    }
  >
    <Svg
      height="16"
      preserveAspectRatio="xMidYMid meet"
      width="22"
    >
      <Path
        d="M2,2L17,2L20,8L17,14L2,14Z"
        fill="rgb(255, 128, 0)"
        stroke="#fff"
        strokeLinecap="round"
        strokeLinejoin="round"
        strokeWidth={2}
      />
    </Svg>
  </AnimatedComponent>
</View>
`;

exports[`Hue components componentWillReceiveProps render 1`] = `
<View
  style={
    Object {
      "height": 10,
      "width": undefined,
    }
  }
>
  <View
    style={
      Object {
        "borderRadius": NaN,
        "height": 10,
        "overflow": "hidden",
        "width": undefined,
      }
    }
  >
    <LinearGradient
      gradientId="hue-bright-cross"
      stops={
        Object {
          "0%": "#FF0000",
          "100%": "#FF0000",
          "12%": "#F8CB0E",
          "24%": "#80FE06",
          "36%": "#08FB2B",
          "49%": "#04FAFC",
          "60%": "#0243FC",
          "70%": "#8700F9",
          "82%": "#FC00EF",
          "92%": "#F00A5B",
        }
      }
      style={
        Object {
          "height": 10,
          "width": undefined,
        }
      }
      x1="0%"
      x2="0%"
      y1="0%"
      y2="100%"
    >
      <Rect
        height={10}
        rx={0}
        ry={0}
        width={0}
        x={0}
        y={0}
      />
    </LinearGradient>
  </View>
  <View
    style={
      Array [
        Object {
          "alignItems": "flex-start",
          "bottom": 0,
          "justifyContent": "center",
          "position": "absolute",
        },
        Object {
          "bottom": -9,
          "height": 18,
          "left": NaN,
          "transform": Array [
            Object {
              "translateX": -5,
            },
            Object {
              "translateY": -5,
            },
            Object {
              "rotate": "-90deg",
            },
          ],
          "width": 10,
        },
      ]
    }
  >
    <TYText
      align={null}
      color={null}
      size={null}
      style={
        Array [
          Object {
            "fontSize": 12,
          },
          Object {
            "fontSize": 12,
          },
        ]
      }
      type={null}
      weight={null}
    />
  </View>
  <AnimatedComponent
    style={
      Object {
        "position": "absolute",
        "right": NaN,
        "top": -8,
      }
    }
  >
    <Svg
      height="16"
      preserveAspectRatio="xMidYMid meet"
      width="22"
    >
      <Path
        d="M2,2L17,2L20,8L17,14L2,14Z"
        fill="rgb(255, 0, 0)"
        stroke="#fff"
        strokeLinecap="round"
        strokeLinejoin="round"
        strokeWidth={2}
      />
    </Svg>
  </AnimatedComponent>
</View>
`;
